### PR TITLE
EstimatorManagerBase refactor

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

* Refactor correct averaging, unified driver, and crowd support out of EstimatorManagerBase and into EstimatorManagerNew.
* Refactor unified driver functionality out of SimpleFixedNodeBranch since much of it needed to be rewritten and into SFNBranch.
* Templatize BranchIO (I imagine temporarily) to support SimpleFixedNodeBranch or SFNBranch.
* Refactor the tests and test classes for the above classes.
* Adjust build including making the estimators a proper subdirectory and library as well as adding refactor to classes.

I believe this is cleaner and easier to review than the previous mark methods in EMB and SFNB legacy or unified.
When the unified driver work is done EstimatorManagerBase and SimpleFixedNodeBranch can be deleted and the factored to classes renamed if necessary.

## What type(s) of changes does this code introduce?

- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)
- Build related changes

### Does this introduce a breaking change?

- No legacy goes on doing its thing with no change.

## What systems has this change been tested on?

- leconte

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes/No. Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed

### CODE REVIEW

**Comment by PDoakORNL** [comment]
## Current State of this PR vs. short.*density tests

```shell
Test project /home/epd/qmcpack/build_gcc
      Start 183: short-LiH_dimer_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4
      Start 279: short-LiH_dimer_pp-vmc_hf_sdj_estimator_spindensity-4-4
 1/20 Test #279: short-LiH_dimer_pp-vmc_hf_sdj_estimator_spindensity-4-4 ...................***Failed    0.79 sec
      Start 283: short-LiH_dimer_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4
 2/20 Test #283: short-LiH_dimer_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4 .........   Passed    9.46 sec
      Start 712: short-diamondC_1x1x1_pp-vmc-estimator-density-4-4
 3/20 Test #712: short-diamondC_1x1x1_pp-vmc-estimator-density-4-4 .........................   Passed   34.87 sec
      Start 714: short-diamondC_1x1x1_pp-dmc-estimator-density-4-4
 4/20 Test #183: short-LiH_dimer_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4 .........   Passed  445.87 sec
      Start 720: short-diamondC_1x1x1_pp-vmc-estimator-spindensity-4-4
 5/20 Test #720: short-diamondC_1x1x1_pp-vmc-estimator-spindensity-4-4 .....................   Passed   38.50 sec
      Start 722: short-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4
 6/20 Test #714: short-diamondC_1x1x1_pp-dmc-estimator-density-4-4 .........................   Passed  514.10 sec
      Start 736: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4
 7/20 Test #736: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4 ..............   Passed  166.48 sec
      Start 738: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-voronoi-4-4
 8/20 Test #738: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-voronoi-4-4 ...........   Passed  125.62 sec
      Start 740: short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4
 9/20 Test #722: short-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4 .....................   Passed  520.63 sec
      Start 184: short-LiH_dimer_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check
      Start 280: short-LiH_dimer_pp-vmc_hf_sdj_estimator_spindensity-4-4-check
      Start 284: short-LiH_dimer_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check
      Start 713: short-diamondC_1x1x1_pp-vmc-estimator-density-4-4-check
      Start 715: short-diamondC_1x1x1_pp-dmc-estimator-density-4-4-check
      Start 721: short-diamondC_1x1x1_pp-vmc-estimator-spindensity-4-4-check
      Start 723: short-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4-check
      Start 737: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4-check
      Start 739: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-voronoi-4-4-check
10/20 Test #280: short-LiH_dimer_pp-vmc_hf_sdj_estimator_spindensity-4-4-check .............***Failed    1.23 sec
11/20 Test #713: short-diamondC_1x1x1_pp-vmc-estimator-density-4-4-check ...................   Passed    1.80 sec
12/20 Test #284: short-LiH_dimer_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check ...***Failed    2.24 sec
13/20 Test #721: short-diamondC_1x1x1_pp-vmc-estimator-spindensity-4-4-check ...............   Passed    2.53 sec
14/20 Test #739: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-voronoi-4-4-check .....   Passed    2.80 sec
15/20 Test #723: short-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4-check ...............***Failed    2.80 sec
16/20 Test #184: short-LiH_dimer_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check ...***Failed    2.82 sec
17/20 Test #715: short-diamondC_1x1x1_pp-dmc-estimator-density-4-4-check ...................***Failed    2.81 sec
18/20 Test #737: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4-check ........   Passed    3.50 sec
19/20 Test #740: short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4 ..............   Passed  599.11 sec
      Start 741: short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check
20/20 Test #741: short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check ........***Failed    1.81 sec
```

## Current State of Develop
```shell
 1/20 Test #721: short-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4 .....................   Passed  521.63 sec
      Start 182: short-LiH_dimer_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4
 2/20 Test #739: short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4 ..............   Passed  732.58 sec
      Start 735: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4
 3/20 Test #735: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4 ..............   Passed  187.71 sec
      Start 737: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-voronoi-4-4
 4/20 Test #182: short-LiH_dimer_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4 .........   Passed  445.65 sec
      Start 719: short-diamondC_1x1x1_pp-vmc-estimator-spindensity-4-4
 5/20 Test #719: short-diamondC_1x1x1_pp-vmc-estimator-spindensity-4-4 .....................   Passed   37.44 sec
      Start 711: short-diamondC_1x1x1_pp-vmc-estimator-density-4-4
 6/20 Test #711: short-diamondC_1x1x1_pp-vmc-estimator-density-4-4 .........................   Passed   35.11 sec
      Start 282: short-LiH_dimer_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4
 7/20 Test #282: short-LiH_dimer_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4 .........   Passed    9.33 sec
      Start 278: short-LiH_dimer_pp-vmc_hf_sdj_estimator_spindensity-4-4
 8/20 Test #278: short-LiH_dimer_pp-vmc_hf_sdj_estimator_spindensity-4-4 ...................***Failed    0.56 sec
      Start 713: short-diamondC_1x1x1_pp-dmc-estimator-density-4-4
 9/20 Test #737: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-voronoi-4-4 ...........   Passed  151.22 sec
      Start 736: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4-check
      Start 738: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-voronoi-4-4-check
      Start 720: short-diamondC_1x1x1_pp-vmc-estimator-spindensity-4-4-check
      Start 712: short-diamondC_1x1x1_pp-vmc-estimator-density-4-4-check
      Start 183: short-LiH_dimer_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check
      Start 279: short-LiH_dimer_pp-vmc_hf_sdj_estimator_spindensity-4-4-check
      Start 283: short-LiH_dimer_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check
      Start 722: short-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4-check
      Start 740: short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check
10/20 Test #279: short-LiH_dimer_pp-vmc_hf_sdj_estimator_spindensity-4-4-check .............***Failed    1.30 sec
11/20 Test #720: short-diamondC_1x1x1_pp-vmc-estimator-spindensity-4-4-check ...............   Passed    2.07 sec
12/20 Test #712: short-diamondC_1x1x1_pp-vmc-estimator-density-4-4-check ...................   Passed    2.07 sec
13/20 Test #738: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-voronoi-4-4-check .....***Failed    2.87 sec
14/20 Test #183: short-LiH_dimer_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check ...***Failed    3.27 sec
15/20 Test #283: short-LiH_dimer_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check ...***Failed    3.38 sec
16/20 Test #722: short-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4-check ...............***Failed    3.38 sec
17/20 Test #740: short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check ........***Failed    3.38 sec
18/20 Test #736: short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4-check ........***Failed    3.71 sec
19/20 Test #713: short-diamondC_1x1x1_pp-dmc-estimator-density-4-4 .........................   Passed  528.27 sec
      Start 714: short-diamondC_1x1x1_pp-dmc-estimator-density-4-4-check
20/20 Test #714: short-diamondC_1x1x1_pp-dmc-estimator-density-4-4-check ...................***Failed    1.44 sec
```

**Comment by prckent** [comment]
Lets consider merging after seeing updated test results. If they are no worse, presumably this can be merged?

**Comment by jtkrogel** [comment]
Mind posting the `--verbose` output for the following tests:
```
short-diamondC_1x1x1_pp-vmc-estimator-density-4-4
short-diamondC_1x1x1_pp-vmc-estimator-spindensity-4-4

short-diamondC_1x1x1_pp-dmc-estimator-density-4-4
short-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4

short-LiH_dimer_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4
short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4
short-diamondC_1x1x1_pp-vmc-estimator-energydensity-voronoi-4-4
```

I would like to check that the deterministic portions (strict checksums) have passed for each of these.

**Comment by ye-luo** [comment]
Review under way.

File: src/Message/MPIObjectBase.h [hunk 1]
```diff
@@ -40,6 +40,9 @@ class MPIObjectBase
   ///return myComm
   inline Communicate* getCommunicator() const { return myComm; }
 
+  ///return a TEMPORARY reference to Communicate
+  inline Communicate& getCommRef() const { return *myComm; }
```
Comment (ye-luo) [inline src/Message/MPIObjectBase.h:44]: Like this being added and encouraged whenever appropriate.

File: src/Estimators/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+IF(USE_OBJECT_TARGET)
+  ADD_LIBRARY(qmcestimators OBJECT ${QMCEST_SRC})
+  ADD_LIBRARY(qmcestimators_unit OBJECT ${QMCEST_SRC})
+ELSE()
+  ADD_LIBRARY(qmcestimators ${QMCEST_SRC})
+  ADD_LIBRARY(qmcestimators_unit ${QMCEST_SRC})
+ENDIF()
+USE_FAKE_RNG(qmcestimators_unit)
+
+TARGET_INCLUDE_DIRECTORIES(qmcestimators PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}")
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/Estimators/CMakeLists.txt:35]: Done

File: src/QMCDrivers/CMakeLists.txt [hunk 3]
```diff
@@ -107,9 +101,9 @@ IF(BUILD_LMYENGINE_INTERFACE)
      )
 ENDIF(BUILD_LMYENGINE_INTERFACE)
 
-SET(QMCDRIVERS ${QMCDRIVERS}
-  ../Estimators/CSEnergyEstimator.cpp
-)
+# SET(QMCDRIVERS ${QMCDRIVERS}
+#   ../Estimators/CSEnergyEstimator.cpp
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCDrivers/CMakeLists.txt:105]: It gets build in Estimators

**Comment by PDoakORNL** [comment]
Results from the density tests. 
https://www.dropbox.com/sh/owjsaroqwnppzag/AADo9wJEjq15r_Vb2KsXn6G7a?dl=0

I think excepting some stochastic effects these are the same, but the signal to noise is pretty low.

**Comment by jtkrogel** [comment]
Thanks Peter.  These all look good.  There are no deterministic changes and the statistical ones show the same properties.

File: src/QMCDrivers/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/CMakeLists.txt:105]: Can we remove the commented lines?

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCDrivers/CMakeLists.txt]: good idea, done

**Comment by prckent** [comment]
Test this please

**Comment by ye-luo** [comment]
test this please

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**